### DISCUSSION TITLE
Support for cword in custom function?

### DISCUSSION THREAD

**Discussion by marcin-krajewski** [JetBrains/ideavim#723] [Q&A] [answered]
Hi
I would like to write custom function to swap some values, for example swap `true -> false`, `false -> true` and so on.
I have function defined like this:
```
function! SwapBoolean()
    let s:w = expand("<cword>")
    if s:w == "true"
      normal ciwfalse
    elseif s:w == "false"
      normal ciwtrue
    endif
endfunction
```
and defined mapping as:
```
" swap true <-> false
nnoremap <Leader><Leader>sw :call SwapBoolean()<CR>
```
Unfortunately when I run it I get the following error:
```
VIM - E121: Undefined variable: s:w
```
As far as I understand `expand` is not supported

Is there any option to get value of current word in custom function? Even if my problem can be simplified and achieved in different way I would like to know if that's possible as I have some other ideas in my mind to reuse that very useful possibility to get current word.

Thanks in advance

**Comment by lippfi** [comment, marked as answer]
Hello!

I attempted to create the same mapping a year ago, and this is what I had in my .ideavimrc (it's a bit rough, but you can grasp the general idea). However, a lot has changed over the year, and there might now be additional functions that could facilitate a smoother implementation (e.g., utilizing dictionary-related functions instead of lists).
```
function! Invert(calledFromVisual)
  let antonyms = ['true', 'false', 'after', 'before', 'start', 'end', 'left', 'right', 'first', 'last',
                 \'True', 'False', 'After', 'Before', 'Start', 'End', 'Left', 'Right', 'First', 'Last']

  if a:calledFromVisual
    normal gv"wy
  else
    normal "wyiw
  endif
  let wordUnderCaret = @w

  let eraseWord = a:calledFromVisual ? 'gvc' : 'ciw'
  let count = 0
  while (count < len(antonyms))
    if (antonyms[count] ==# wordUnderCaret)
      let antonym = (count % 2 ==? 0) ? antonyms[count + 1] : antonyms[count - 1]
      execute 'normal ' .. eraseWord .. antonym
      break
    endif
    let count += 1
  endwhile
endfunction

nnoremap ! :call Invert(0)<CR>
vnoremap ! :<C-u>call Invert(1)<CR>
```

IdeaVim is an open-source project, and any PRs with new functions are highly appreciated. You are very welcome to create a PR with any functions you need. Currently, the IdeaVim maintainers are focused on bug fixes and Fleet IDE support, so we are unable to add the expand function or any other function at the moment. You can find the list of supported functions here:

 [functions part 1](https://github.com/JetBrains/ideavim/blob/master/vim-engine/src/main/resources/engine_vimscript_functions.json),
 [functions part 2](https://github.com/JetBrains/ideavim/blob/master/src/main/resources/intellij_vimscript_functions.json)

I also recommend writing plugins in Kotlin. That's the reason I stopped using the script I wrote a year ago. IDE integration is what makes IdeaVim so powerful. If you create your own plugin for the IDE, you can invoke its actions via the `<Action>` mappings. IDE plugins written in Kotlin will provide you with more capabilities, such as access to PSI. For instance, not only will it enable you to swap true and false, but it will also allow you to cycle through enum values or execute other advanced, language-aware operations.

**Reply by marcin-krajewski** [reply]
Thanks a lot for your response. Applying concept from your code helped me to achieve what I need.

Writing functions in kotlin will be possible only in case that's IDE specific use case. Like eg. built-in Intellij IDE functions, NextSplitter, ReformatCode and so on. Otherwise I need (want!) to be consistent with pure VIM as I'm using VIM and IDE 50% - 50%. When I will decide to use kotlin it will make the function useless in VIM.
